File: src/forimpatient/chapter04/Exercise09.scala
package forimpatient.chapter04

/**
  * Created by Iryna Kharaborkina on 7/30/16. 
  *
  * Solution to the Chapter 04 Exercise 09 'Scala for the Impatient' by Horstmann C.S.
  *
  * Write a function lteqgt(values: Array[Int], v: Int) that returns a triple containing
  * the counts of values less than v, equal to v, and greater than v.
  */
object Exercise09 extends App {
  println("Chapter 04 Exercise 09")

  println(lteqgt((1 to 10).toArray, 4))
  println(lteqgt(Array(5, 7, -1,  9, 2, -6), 2))

  def lteqgt(values: Array[Int], v: Int) = (values.count(_ < v), values.count(_ == v), values.count(_ > v))
}

File: src/forimpatient/chapter05/Exercise09.java
package forimpatient.chapter05;

/**
 * Created by Iryna Kharaborkina on 8/3/16.
 * <p>
 * Solution to the Chapter 05 Exercise 09 'Scala for the Impatient' by Horstmann C.S.
 *
 *
 */
class Exercise09 {
    public static void main(String[] args) {
        println("Chapter 05 Exercise 09");

        Car a = new Car("tesla", "fdf445", 2015, "dfdfdf");

        println(a.getManufacturer());
        println(a.getModelName());
        println(a.getModelYear());
        println(a.getLicense());

        a.setLicense("retrtr");
        println(a.getLicense());

        Car b = new Car("ford", "fdsfds");

        println(b.getManufacturer());
        println(b.getModelName());
        println(b.getModelYear());
        println(b.getLicense());

        b.setLicense("tytyty");
        println(b.getLicense());
    }

    static class Car {
        private final String manufacturer;
        private final String modelName;
        private final int modelYear;
        private String license;

        Car(String manufacturer, String modelName, int modelYear, String license) {
            this.manufacturer = manufacturer;
            this.modelName = modelName;
            this.modelYear = modelYear;
            this.license = license;
        }

        Car(String manufacturer, String modelName, int modelYear) {
            this(manufacturer, modelName, modelYear, "");
        }

        Car(String manufacturer, String modelName, String license) {
            this(manufacturer, modelName, -1, license);
        }

        Car(String manufacturer, String modelName) {
            this(manufacturer, modelName, -1, "");
        }

        String getManufacturer() {
            return manufacturer;
        }

        String getModelName() {
            return modelName;
        }

        int getModelYear() {
            return modelYear;
        }

        String getLicense() {
            return license;
        }

        void setLicense(String license) {
            this.license = license;
        }
    }

    private static void println(Object txt) {
        System.out.println(txt);
    }
}
